cpp/node_similarity_module/algorithms/node_similarity.hpp

```diff
@@ -0,0 +1,193 @@
+}
+
+/*
+Calculates overlap function between two set of neighbours.
+ns1 are neighbours of the first node.
+ns2 are neighbours of the second node.
+*/
+double OverlapFunc(const std::set<uint64_t> &ns1, const std::set<uint64_t> &ns2) {
+    std::set<uint64_t> elem_intersection;
+    std::set_intersection(ns1.begin(), ns1.end(), ns2.begin(), ns2.end(), std::inserter(elem_intersection, elem_intersection.begin()));
+    int denominator = std::min(ns1.size(), ns2.size());
+    if (denominator == 0) {
+        return 0.0;
+    }
+    return elem_intersection.size() / (double) denominator;
+}
+
+
+double CosineFunc(const mgp::List &prop1, const mgp::List &prop2) {
+    int size = prop1.Size();
+    double similarity = 0.0, node1_sum = 0.0, node2_sum = 0.0;
+    for (int i = 0; i < size; ++i) {
+        double val1 = prop1[i].ValueDouble(), val2 = prop2[i].ValueDouble();
+        similarity += val1 * val2;
+        node1_sum += val1 * val1;
+        node2_sum += val2 * val2;
+    }
+    double denominator = sqrt(node1_sum) * sqrt(node2_sum);
+    if (denominator < 1e-9) {
+        return 0.0;
+    } 
+    return similarity / denominator;
+
+}
+
+
+
+/*
+Calculates cosine similarity function between two nodes for a given property.
+*/
+double CosineFuncWrapper(const mgp::Node &node1, const mgp::Node &node2, const std::string &property) {
+    const auto &prop1_it = node1.GetProperty(property);
+    const auto &prop2_it = node2.GetProperty(property);
+    if (prop1_it.IsNull() || prop2_it.IsNull()) {
+        throw mgp::ValueException("All nodes should have property " + property);
+    }
+    const auto &prop1 = prop1_it.ValueList();
+    const auto &prop2 = prop2_it.ValueList();
+    int size1 = prop1.Size(), size2 = prop2.Size();
+    if (size1 != size2) {
+        throw mgp::ValueException("Vectors should be of the same size.");
+    }
+    return CosineFunc(prop1, prop2);
+}
+
+
+
+/*
+Extract node neighbours.
+*/
+std::set<uint64_t> GetNeighbors(std::unordered_map<uint64_t, std::set<uint64_t>> &all_node_neighbors, const mgp::Node &node) {
+    uint64_t node_id = node.Id().AsUint();
+    const auto &result_it = all_node_neighbors.find(node_id);
+    if (result_it != all_node_neighbors.end()) {
+        return result_it->second;
+    }
+    all_node_neighbors[node_id] = std::set<uint64_t>();
+    auto &ns = all_node_neighbors[node_id];
+    const auto &rels = node.OutRelationships();
+    for (const auto &rel: rels) {
+        ns.insert(rel.To().Id().AsUint());
+    }
+    return ns;
+}
+
+
+/*
+Calculates similiraty between pairs of nodes given by src_nodes and dst_nodes.
+*/
+std::vector<std::tuple<mgp::Node, mgp::Node, double>> CalculateSimilarityPairwise(const mgp::List &src_nodes, const mgp::List &dst_nodes, node_similarity_util::Similarity similarity_mode, const std::string &property = "") {
+    if (src_nodes.Size() != dst_nodes.Size()) {
+        throw mgp::ValueException("Arguments are of different size.");
```

Comment:
More helpful this way
```suggestion
        throw mgp::ValueException("The node lists must be the same length.");
```